File: core/guardrails.py
"""Guardrails — Parallel safety checks for agent execution.

Inspired by OpenAI Agents SDK's guardrail pattern, guardrails run
in parallel with the main execution flow and can:
- PASS: Allow the execution to continue
- WARN: Log a warning but allow execution
- BLOCK: Stop execution immediately
- MODIFY: Transform the input/output before proceeding

Usage:
    from flowforge.core.guardrails import InputGuardrail, OutputGuardrail, GuardrailResult

    class ContentSafetyGuardrail(InputGuardrail):
        name = "content_safety"

        async def check(self, input_text: str, context: dict) -> GuardrailResult:
            if any(word in input_text for word in BANNED_WORDS):
                return GuardrailResult(status="blocked", message="Contains banned content")
            return GuardrailResult(status="passed")

    class QualityGuardrail(OutputGuardrail):
        name = "quality_check"

        async def check(self, output_text: str, context: dict) -> GuardrailResult:
            if len(output_text) < 100:
                return GuardrailResult(status="warned", message="Output too short")
            return GuardrailResult(status="passed")
"""

from __future__ import annotations

import asyncio
from abc import ABC, abstractmethod
from typing import Literal, Optional

from pydantic import BaseModel, Field

from flowforge.core.tracing import get_logger

logger = get_logger("guardrails")


class GuardrailResult(BaseModel):
    """Result of a guardrail check.

    Attributes:
        status: One of "passed", "warned", "blocked", "modified".
        message: Human-readable description of the result.
        modified_data: When status is "modified", contains the transformed data.
    """

    status: Literal["passed", "warned", "blocked", "modified"] = "passed"
    message: str = ""
    modified_data: Optional[str] = None


class InputGuardrail(ABC):
    """Base class for input guardrails.

    Subclasses must implement the ``check`` method which inspects
    the input text and returns a GuardrailResult.
    """

    name: str = "unnamed_input_guardrail"

    @abstractmethod
    async def check(self, input_text: str, context: dict) -> GuardrailResult:
        """Check the input text against this guardrail's rules.

        Args:
            input_text: The text to inspect.
            context: Additional context from the execution environment.

        Returns:
            A GuardrailResult indicating pass/warn/block/modify.
        """
        ...


class OutputGuardrail(ABC):
    """Base class for output guardrails.

    Subclasses must implement the ``check`` method which inspects
    the output text and returns a GuardrailResult.
    """

    name: str = "unnamed_output_guardrail"

    @abstractmethod
    async def check(self, output_text: str, context: dict) -> GuardrailResult:
        """Check the output text against this guardrail's rules.

        Args:
            output_text: The text to inspect.
            context: Additional context from the execution environment.

        Returns:
            A GuardrailResult indicating pass/warn/block/modify.
        """
        ...


class GuardrailRegistry:
    """Registry for managing input and output guardrails.

    Guardrails are registered separately as input or output checks
    and can be looked up by name.
    """

    def __init__(self) -> None:
        self._input_guardrails: dict[str, InputGuardrail] = {}
        self._output_guardrails: dict[str, OutputGuardrail] = {}

    def register(self, guardrail: InputGuardrail | OutputGuardrail) -> None:
        """Register a guardrail instance.

        InputGuardrail instances are stored in the input registry;
        OutputGuardrail instances are stored in the output registry.

        Args:
            guardrail: An InputGuardrail or OutputGuardrail instance.
        """
        if isinstance(guardrail, InputGuardrail):
            if guardrail.name in self._input_guardrails:
                logger.debug(f"Input guardrail '{guardrail.name}' already registered, skipping duplicate")
                return
            self._input_guardrails[guardrail.name] = guardrail
            logger.info(f"Registered input guardrail: {guardrail.name}")
        elif isinstance(guardrail, OutputGuardrail):
            if guardrail.name in self._output_guardrails:
                logger.debug(f"Output guardrail '{guardrail.name}' already registered, skipping duplicate")
                return
            self._output_guardrails[guardrail.name] = guardrail
            logger.info(f"Registered output guardrail: {guardrail.name}")
        else:
            raise TypeError(f"Expected InputGuardrail or OutputGuardrail, got {type(guardrail)}")

    def unregister(self, name: str) -> None:
        """Remove a guardrail by name from both input and output registries.

        Args:
            name: The guardrail name to unregister.

        Raises:
            KeyError: If the name is not found in either registry.
        """
        found = False
        if name in self._input_guardrails:
            del self._input_guardrails[name]
            found = True
        if name in self._output_guardrails:
            del self._output_guardrails[name]
            found = True
        if not found:
            raise KeyError(f"Guardrail '{name}' not registered")

    def get_input_guardrails(self) -> list[InputGuardrail]:
        """Return all registered input guardrails."""
        return list(self._input_guardrails.values())

    def get_output_guardrails(self) -> list[OutputGuardrail]:
        """Return all registered output guardrails."""
        return list(self._output_guardrails.values())


class GuardrailExecutor:
    """Executes guardrails in parallel using asyncio.gather.

    If any guardrail returns "blocked", execution stops immediately
    and the blocked result is returned along with results collected
    so far.
    """

    def __init__(self, registry: GuardrailRegistry) -> None:
        self._registry = registry

    async def run_input_guardrails(
        self, input_text: str, context: dict
    ) -> list[GuardrailResult]:
        """Run all input guardrails in parallel.

        Args:
            input_text: The input text to check.
            context: Additional execution context.

        Returns:
            A list of GuardrailResult from each guardrail.
            If any result is "blocked", returns immediately with
            results collected so far plus the blocked result.
        """
        guardrails = self._registry.get_input_guardrails()
        if not guardrails:
            return []

        tasks = [g.check(input_text, context) for g in guardrails]
        results = await self._run_parallel(tasks, guardrails)
        return results

    async def run_output_guardrails(
        self, output_text: str, context: dict
    ) -> list[GuardrailResult]:
        """Run all output guardrails in parallel.

        Args:
            output_text: The output text to check.
            context: Additional execution context.

        Returns:
            A list of GuardrailResult from each guardrail.
            If any result is "blocked", returns immediately with
            results collected so far plus the blocked result.
        """
        guardrails = self._registry.get_output_guardrails()
        if not guardrails:
            return []

        tasks = [g.check(output_text, context) for g in guardrails]
        results = await self._run_parallel(tasks, guardrails)
        return results

    async def _run_parallel(
        self,
        tasks: list,
        guardrails: list[InputGuardrail | OutputGuardrail],
    ) -> list[GuardrailResult]:
        """Run guardrail check tasks in parallel with early blocking.

        Uses asyncio.gather with return_exceptions=True so that one
        failing guardrail does not prevent others from completing.
        If any guardrail returns "blocked", we return immediately.
        """
        raw_results = await asyncio.gather(*tasks, return_exceptions=True)

        results: list[GuardrailResult] = []
        for i, raw in enumerate(raw_results):
            guardrail_name = guardrails[i].name
            if isinstance(raw, Exception):
                logger.error(f"Guardrail '{guardrail_name}' raised an error: {raw}")
                results.append(
                    GuardrailResult(
                        status="warned",
                        message=f"Guardrail '{guardrail_name}' error: {raw}",
                    )
                )
            elif isinstance(raw, GuardrailResult):
                results.append(raw)
                if raw.status == "blocked":
                    logger.warning(f"Guardrail '{guardrail_name}' blocked execution: {raw.message}")
                    return results
                elif raw.status == "warned":
                    logger.warning(f"Guardrail '{guardrail_name}' warning: {raw.message}")
                elif raw.status == "modified":
                    logger.info(f"Guardrail '{guardrail_name}' modified data: {raw.message}")
            else:
                logger.error(f"Guardrail '{guardrail_name}' returned unexpected type: {type(raw)}")
                results.append(
                    GuardrailResult(
                        status="warned",
                        message=f"Guardrail '{guardrail_name}' returned unexpected type",
                    )
                )

        return results

File: packages/plugins/guard/src/guardrails.ts
/**
 * Guardrails — parallel safety checks for agent execution.
 *
 * Maps flowforge Python legacy core/guardrails.py (F25). Guardrails run
 * in parallel with the main execution flow and can:
 * - PASS: allow execution to continue
 * - WARN: log a warning but allow execution
 * - BLOCK: stop execution immediately
 * - MODIFY: transform the input/output before proceeding
 */

export type GuardrailStatus = 'passed' | 'warned' | 'blocked' | 'modified'

export interface GuardrailResultInit {
  status?: GuardrailStatus
  message?: string
  /** Present when status is 'modified': the transformed data. */
  modifiedData?: string | null
}

export class GuardrailResult {
  readonly status: GuardrailStatus
  readonly message: string
  readonly modifiedData: string | null

  constructor(init: GuardrailResultInit = {}) {
    this.status = init.status ?? 'passed'
    this.message = init.message ?? ''
    this.modifiedData = init.modifiedData ?? null
  }
}

export type GuardrailContext = Record<string, unknown>

export abstract class InputGuardrail {
  name = 'unnamed_input_guardrail'
  abstract check(inputText: string, context: GuardrailContext): Promise<GuardrailResult>
}

export abstract class OutputGuardrail {
  name = 'unnamed_output_guardrail'
  abstract check(outputText: string, context: GuardrailContext): Promise<GuardrailResult>
}

export type Guardrail = InputGuardrail | OutputGuardrail

export interface GuardrailLogger {
  info(message: string): void
  warning(message: string): void
  error(message: string): void
}

const noopLogger: GuardrailLogger = {
  info: () => {},
  warning: () => {},
  error: () => {},
}

export class GuardrailRegistry {
  private readonly inputGuardrails = new Map<string, InputGuardrail>()
  private readonly outputGuardrails = new Map<string, OutputGuardrail>()

  constructor(private readonly logger: GuardrailLogger = noopLogger) {}

  /** Register a guardrail; duplicates (same name + side) are skipped. */
  register(guardrail: Guardrail): void {
    if (guardrail instanceof InputGuardrail) {
      if (this.inputGuardrails.has(guardrail.name)) {
        this.logger.info(`Input guardrail '${guardrail.name}' already registered, skipping duplicate`)
        return
      }
      this.inputGuardrails.set(guardrail.name, guardrail)
      this.logger.info(`Registered input guardrail: ${guardrail.name}`)
    } else if (guardrail instanceof OutputGuardrail) {
      if (this.outputGuardrails.has(guardrail.name)) {
        this.logger.info(`Output guardrail '${guardrail.name}' already registered, skipping duplicate`)
        return
      }
      this.outputGuardrails.set(guardrail.name, guardrail)
      this.logger.info(`Registered output guardrail: ${guardrail.name}`)
    } else {
      throw new TypeError(
        `Expected InputGuardrail or OutputGuardrail, got ${String(guardrail)}`,
      )
    }
  }

  /** Remove a guardrail by name from both registries; throws when absent. */
  unregister(name: string): void {
    let found = false
    if (this.inputGuardrails.delete(name)) found = true
    if (this.outputGuardrails.delete(name)) found = true
    if (!found) throw new Error(`Guardrail '${name}' not registered`)
  }

  getInputGuardrails(): InputGuardrail[] {
    return [...this.inputGuardrails.values()]
  }

  getOutputGuardrails(): OutputGuardrail[] {
    return [...this.outputGuardrails.values()]
  }
}

/**
 * Executes guardrails in parallel (Promise.allSettled semantics): one
 * failing guardrail never prevents the others from completing. A
 * 'blocked' result stops collection immediately, returning the results
 * gathered so far including the blocking one.
 */
export class GuardrailExecutor {
  constructor(
    private readonly registry: GuardrailRegistry,
    private readonly logger: GuardrailLogger = noopLogger,
  ) {}

  async runInputGuardrails(
    inputText: string,
    context: GuardrailContext,
  ): Promise<GuardrailResult[]> {
    const guardrails = this.registry.getInputGuardrails()
    if (guardrails.length === 0) return []
    const tasks = guardrails.map(guardrail => guardrail.check(inputText, context))
    return this.runParallel(tasks, guardrails)
  }

  async runOutputGuardrails(
    outputText: string,
    context: GuardrailContext,
  ): Promise<GuardrailResult[]> {
    const guardrails = this.registry.getOutputGuardrails()
    if (guardrails.length === 0) return []
    const tasks = guardrails.map(guardrail => guardrail.check(outputText, context))
    return this.runParallel(tasks, guardrails)
  }

  private async runParallel(
    tasks: Array<Promise<GuardrailResult>>,
    guardrails: Guardrail[],
  ): Promise<GuardrailResult[]> {
    const settled = await Promise.allSettled(tasks)
    const results: GuardrailResult[] = []
    for (let i = 0; i < settled.length; i++) {
      const outcome = settled[i] as PromiseSettledResult<GuardrailResult>
      const guardrailName = guardrails[i]?.name ?? `guardrail-${i}`
      if (outcome.status === 'rejected') {
        this.logger.error(`Guardrail '${guardrailName}' raised an error: ${outcome.reason}`)
        results.push(
          new GuardrailResult({
            status: 'warned',
            message: `Guardrail '${guardrailName}' error: ${outcome.reason}`,
          }),
        )
        continue
      }
      const raw = outcome.value
      if (raw instanceof GuardrailResult) {
        results.push(raw)
        if (raw.status === 'blocked') {
          this.logger.warning(
            `Guardrail '${guardrailName}' blocked execution: ${raw.message}`,
          )
          return results
        }
        if (raw.status === 'warned') {
          this.logger.warning(`Guardrail '${guardrailName}' warning: ${raw.message}`)
        } else if (raw.status === 'modified') {
          this.logger.info(`Guardrail '${guardrailName}' modified data: ${raw.message}`)
        }
      } else {
        this.logger.error(
          `Guardrail '${guardrailName}' returned unexpected type: ${typeof raw}`,
        )
        results.push(
          new GuardrailResult({
            status: 'warned',
            message: `Guardrail '${guardrailName}' returned unexpected type`,
          }),
        )
      }
    }
    return results
  }
}
